### PR TITLE
feat: Integrate Authorization Filters in Search Query

### LINKED ISSUE #23174: Integrate Authorization Filters in Search Query
## Description

Extend search queries to include authorization filters based on the authenticated user’s context (userKey, groups ids, and roles ids) + required permission checks passed to the search client.

* In the document based implementation, fetch the authorized resource keys from the`zeebe-record-authorization` index based on the following query:
```
(auth.ownerType:=USER AND auth.ownerKey:=userKey)
OR (auth.ownerType:=GROUP AND auth.ownerKey IN groupsIds)
OR (auth.ownerType:=ROLE AND auth.ownerKey IN roleIds)
AND auth.resourceType:=resourceType 
AND auth.permissions CONTAINS permissionType
```

* Compute necessary additional filters based on the query type and required permission.
* Example mappings for query types and corresponding authorization details:

| Query type  | auth resource type | auth permission type | filter on field |
| ------------- | ----------------| ----------------------- | ------------- |
| ProcessInstanceQuery | PROCESS_INSTANCE | READ | "key" |
| ProcessInstanceQuery | PROCESS_DEFINITION | READ_INSTANCE | "bpmnProcessId" |

**Note:** The PROCESS_INSTANCE.READ permission is outside the scope of version 8.7, but is included here for illustrative purposes.

* Implement handling for wildcards in authorized resource keys.
* if no authorized resource key is returned, append a `noneMatch` filter to the query
* Ensure that all authorized resourceIds can be retrieved using scrolling.
* Enhance the authorization filter to include permission-based filtering (currently not supported).

| Resource           | Permission | Query   |
|----------------|---------------|---------------|
| Process Definition   | READ  | <ul><li>[ ] Process Definition Query</li></ul> |
| Process Definition   | READ_INSTANCE  | <ul><li>[ ] Process instance Query</li><li>[ ] Flownode instance Query</li><li>[ ] Variable Query</li><li>[ ] Incident Query</li></ul> |
| Process Definition   | READ_TASK  | <ul><li>[ ] User task Query</li></ul> |
| Decision Requirements Definition   | READ  | <ul><li>[ ] Decision Requirements Definition Query</li></ul> |
| Decision Definition   | READ  | <ul><li>[ ] Decision Definition Query</li></ul> |
| Decision Definition   | READ_INSTANCE  | <ul><li>[ ] Decision Instance Query</li></ul> |
| User  | READ  | <ul><li>[ ] User Query</li></ul> |
| Authorization  | READ  | <ul><li>[ ] Authorization Query</li></ul> |

### PR DESCRIPTION
**Pull request by houssain-barouni**
## Description

Adds authorization filters to the search query.
It is based on `DocumentAuthorizationQueryStrategy` that applies authorization to a search query by fetching the authorized resources for the authenticated user and creating a new search query with the authorization applied.

**Remaining:**
- [x] apply scrolling when fetching authorized resource keys: it will be handled in a second pull request: https://github.com/camunda/camunda/pull/24085
- [x] Adding integration tests  for all resource types as `ProcessAuthorizationIT` after validating the testing concept described below

`ProcessAuthorizationIT` integration test is based on `CamundaExporterITInvocationProvider` extension. The extension is registered programitically with `@RegisterExtension` to allow customizable test setup. 

```java
  @RegisterExtension
  static final CamundaExporterITInvocationProvider PROVIDER =
      new CamundaExporterITInvocationProvider()
          .withAdditionalProfiles(Profile.AUTH_BASIC)
          .withAuthorizationsEnabled()
          .withUsers(ADMIN_USER, RESTRICTED_USER);
```

It was also extended to automatically create `ZeebeClient` with user credentials with `@Authenticated` annotation
```java
@TestTemplate
  void shouldReturnAuthorizedProcessDefinitions(
      @Authenticated(ADMIN) final ZeebeClient adminClient,
      @Authenticated(RESTRICTED) final ZeebeClient userClient) {
```

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes https://github.com/camunda/camunda/issues/23174

### CODE REVIEW

File: service/src/main/java/io/camunda/service/DecisionRequirementsServices.java [hunk 1]
```diff
@@ -48,7 +50,16 @@ public DecisionRequirementsServices withAuthentication(final Authentication auth
   public SearchQueryResult<DecisionRequirementsEntity> search(
       final DecisionRequirementsQuery query) {
     return decisionRequirementSearchClient.searchDecisionRequirements(
-        query, SecurityContext.of(s -> s.withAuthentication(authentication)));
+        query,
+        SecurityContext.of(
```
Comment (nathansandi) [inline service/src/main/java/io/camunda/service/DecisionRequirementsServices.java:54]: 💭  [Suggestion/Optional]  I see this is used over all the services, and it is a bit long, would be the case to extract this to a utility class and reuse? 

for example: 
```
public class SecurityContextUtil {

    public static SecurityContext buildSecurityContext(Authentication authentication,
                                                      SecurityConfiguration securityConfig,
                                                      AuthorizationResourceType resourceType,
                                                      PermissionType permissionType) {
        return SecurityContext.of(
            s -> s.withAuthentication(authentication)
                  .withAuthorizationIfEnabled(
                      securityConfig.getAuthorizations().isEnabled(),
                      a -> a.resourceType(resourceType)
                            .permissionType(permissionType)
                  )
        );
    }
}
```

Dont need to be done in this PR, its maybe an idea.

File: security/security-core/src/main/java/io/camunda/zeebe/protocol/record/value/PermissionType.java [hunk 2]
```diff
@@ -11,6 +11,7 @@ public enum PermissionType {
   CREATE,
   READ,
```
Comment (nathansandi) [inline security/security-core/src/main/java/io/camunda/zeebe/protocol/record/value/PermissionType.java:12]: [Question] I assuming we have 3 different reads to this case, once the permission to read instance, may differ to user task, as well a generic one. So it is not possible have only READ here.  Correct?

**Review by nathansandi — APPROVED** [review]
From code wise perspective, on my side looks good!
Left minor suggestion, but optional so far. 

Related to the all requirements I dont feel 100% confident to review, so I will let Ben does that

Great work here :)

File: service/src/main/java/io/camunda/service/DecisionRequirementsServices.java (hunk 1 above)
Comment (houssain-barouni) [reply service/src/main/java/io/camunda/service/DecisionRequirementsServices.java:54]: I thought of doing that too, I will see if I can do this when implementing permission checks on the Get by key endpoints

File: security/security-core/src/main/java/io/camunda/zeebe/protocol/record/value/PermissionType.java (hunk 2 above)
Comment (houssain-barouni) [reply security/security-core/src/main/java/io/camunda/zeebe/protocol/record/value/PermissionType.java:12]: Yes, ProcesDef.READ allows you to read the process definitions but not the instance objects and the user tasks.
This was specified in https://docs.google.com/document/d/1ikxw9ARCv-MBwASnj7WhW5uHCndCRk2B35xi-cB2N0M/edit?tab=t.0#heading=h.m2zg2f7pb01i

**Review by ralfpuchert — APPROVED** [review]

**Pull request merged**